File: apps/mobile/src/lib/nutrition/logic/verdict-logic.ts
/**
 * Pure verdict helpers extracted from web `components/nutrition/sections/
 * verdict-hero.tsx` (keep in sync). The headline branch selection stays in the
 * component (it needs the translator); these are the locale/data-only pieces.
 */
import type {
  NutrientSummaryItem,
  NutritionOverview,
  NutritionRange,
} from '@/lib/nutrition/types';

export const PERIOD_KEYS: Record<NutritionRange, string> = {
  '7d': 'verdict.period7d',
  '30d': 'verdict.period30d',
  '90d': 'verdict.period90d',
};

/** Translator shape — matches use-intl's `useTranslations()` call signature. */
type Translator = (key: string) => string;

export function takeNames(
  items: NutrientSummaryItem[],
  tRoot: Translator,
  limit = 2
): string[] {
  return items
    .slice(0, limit)
    .map((item) => tRoot(item.labelKey).toLowerCase());
}

export function joinNames(names: string[], locale: string): string {
  if (names.length === 0) return '';
  try {
    return new Intl.ListFormat(locale, {
      style: 'long',
      type: 'conjunction',
    }).format(names);
  } catch {
    // Hermes builds without Intl.ListFormat fall back to a plain join.
    return names.join(', ');
  }
}

export function getAverageConfidence(overview: NutritionOverview): number | null {
  const items = [
    ...overview.summary.mostConsistent,
    ...overview.summary.needsAttention,
  ];
  if (items.length === 0) return null;
  return (
    items.reduce((total, item) => total + item.confidence, 0) / items.length
  );
}

File: apps/mobile-flutter/lib/features/nutrition/logic/verdict_logic.dart
/// Pure verdict helpers. The headline branch selection stays in the widget (it
/// needs the translator); these are the locale/data-only pieces.
///
/// No live web counterpart: the `verdict-hero.tsx` section these were
/// extracted from was removed in the web nutrition rewrite (commit cdd7a3fa).
library;

import '../../../models/nutrition/nutrition.dart';

/// `resolvedRange` is a raw '7d' | '30d' | '90d' string in the Dart model.
const Map<String, String> kPeriodKeys = {
  '7d': 'verdict.period7d',
  '30d': 'verdict.period30d',
  '90d': 'verdict.period90d',
};

/// Translator shape — matches easy_localization's `tr(key)` call signature.
typedef Translator = String Function(String key);

List<String> takeNames(
  List<NutrientSummaryItem> items,
  Translator tRoot, {
  int limit = 2,
}) {
  return items
      .take(limit)
      .map((item) => tRoot(item.labelKey).toLowerCase())
      .toList();
}

/// Joins names with a locale-aware conjunction. Dart's `intl` has no
/// `ListFormat`, so this mirrors the RN Hermes fallback: English uses an Oxford
/// "and", Vietnamese uses "và"; everything else falls back to a comma join.
String joinNames(List<String> names, String locale) {
  if (names.isEmpty) return '';
  if (names.length == 1) return names.first;
  final conj = locale.startsWith('vi') ? 'và' : 'and';
  if (names.length == 2) return '${names[0]} $conj ${names[1]}';
  final head = names.sublist(0, names.length - 1).join(', ');
  return '$head, $conj ${names.last}';
}
